commit 447413f00860
Author: Xidorn Quan <me@upsuper.org>
Date:   Thu Sep 28 16:19:45 2017 +1000

    Bug 1402094 - Blacklist about:reader from using stylo. r=bholley, a=sledru
    
    MozReview-Commit-ID: 6cG1x1tcOwn
    
    --HG--
    extra : source : 813556d9e389f580d8e6efa9c24825edec952259
---
 dom/base/nsDocument.cpp | 25 ++++++++++++++++++++++++-
 1 file changed, 24 insertions(+), 1 deletion(-)

diff --git dom/base/nsDocument.cpp dom/base/nsDocument.cpp
index 07ef9ef817b7..7f8ebecffa97 100644
--- dom/base/nsDocument.cpp
+++ dom/base/nsDocument.cpp
@@ -13575,6 +13575,29 @@ nsIDocument::ReportHasScrollLinkedEffect()
                                   "ScrollLinkedEffectFound2");
 }
 
+#ifdef MOZ_STYLO
+// URL-based blacklist for stylo.
+static bool
+ShouldUseGeckoBackend(nsIURI* aDocumentURI)
+{
+  if (!aDocumentURI) {
+    return false;
+  }
+  bool isScheme = false;
+  if (NS_SUCCEEDED(aDocumentURI->SchemeIs("about", &isScheme))) {
+    nsAutoCString path;
+    aDocumentURI->GetFilePath(path);
+    // about:reader requires support of :scope pseudo-class so we have
+    // to use Gecko backend for now. See bug 1402094.
+    // This should be fixed by bug 1204818.
+    if (path.EqualsLiteral("reader")) {
+      return true;
+    }
+  }
+  return false;
+}
+#endif // MOZ_STYLO
+
 void
 nsIDocument::UpdateStyleBackendType()
 {
@@ -13593,7 +13616,7 @@ nsIDocument::UpdateStyleBackendType()
     // Note that, since tests can have XUL support, we still need to
     // explicitly exclude XUL documents here.
     if (!nsContentUtils::IsSystemPrincipal(NodePrincipal()) &&
-        !IsXULDocument()) {
+        !IsXULDocument() && !ShouldUseGeckoBackend(mDocumentURI)) {
       mStyleBackendType = StyleBackendType::Servo;
     }
   }
